File: FruitWholesale.Client/src/app/features/supplier-return/supplier-return.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { PaginatedList, PaginationRequest } from '../../core/models/common.model';
import { SaveSupplierReturn, SupplierReturn, SupplierReturnListItem } from '../../core/models/transactions.model';
import { toHttpParams } from '../../core/utils/http-params.util';

@Injectable({ providedIn: 'root' })
export class SupplierReturnService {
  private readonly baseUrl = `${environment.apiUrl}/supplierreturn`;

  constructor(private readonly http: HttpClient) {}

  getPaged(request: PaginationRequest, supplierId?: number | null, fromDate?: string | null, toDate?: string | null) {
    return this.http.get<PaginatedList<SupplierReturnListItem>>(this.baseUrl, {
      params: toHttpParams({ ...request, supplierId, fromDate, toDate })
    });
  }

  getById(id: number): Observable<SupplierReturn> {
    return this.http.get<SupplierReturn>(`${this.baseUrl}/${id}`);
  }

  getNextReferenceNo(): Observable<string> {
    return this.http.get(`${this.baseUrl}/next-reference-no`, { responseType: 'text' });
  }

  create(dto: SaveSupplierReturn): Observable<SupplierReturn> {
    return this.http.post<SupplierReturn>(this.baseUrl, dto);
  }

  update(id: number, dto: SaveSupplierReturn): Observable<SupplierReturn> {
    return this.http.put<SupplierReturn>(`${this.baseUrl}/${id}`, dto);
  }

  delete(id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl}/${id}`);
  }
}

File: FruitWholesale.Mobile/lib/features/supplier_return/supplier_return_service.dart
import '../../core/api/api_client.dart';
import '../../core/models/paginated_list.dart';
import 'supplier_return_models.dart';

class SupplierReturnService {
  final ApiClient _api;

  SupplierReturnService(this._api);

  Future<PaginatedList<SupplierReturnListItem>> getPaged({
    int pageNumber = 1,
    int pageSize = 20,
    String? searchTerm,
    int? supplierId,
    String? fromDate,
    String? toDate,
  }) async {
    final json = await _api.get('/supplierreturn', query: {
      'pageNumber': pageNumber,
      'pageSize': pageSize,
      if (searchTerm != null && searchTerm.isNotEmpty) 'searchTerm': searchTerm,
      if (supplierId != null) 'supplierId': supplierId,
      if (fromDate != null) 'fromDate': fromDate,
      if (toDate != null) 'toDate': toDate,
    });
    return PaginatedList.fromJson(json as Map<String, dynamic>, SupplierReturnListItem.fromJson);
  }

  Future<SupplierReturnDetail> getById(int id) async {
    final json = await _api.get('/supplierreturn/$id');
    return SupplierReturnDetail.fromJson(json as Map<String, dynamic>);
  }

  Future<String> getNextReferenceNo() async {
    final json = await _api.get('/supplierreturn/next-reference-no');
    return json as String;
  }

  Future<SupplierReturnDetail> create(SupplierReturnDetail supplierReturn) async {
    final json = await _api.post('/supplierreturn', body: supplierReturn.toSaveJson());
    return SupplierReturnDetail.fromJson(json as Map<String, dynamic>);
  }

  Future<SupplierReturnDetail> update(int id, SupplierReturnDetail supplierReturn) async {
    final json = await _api.put('/supplierreturn/$id', body: supplierReturn.toSaveJson());
    return SupplierReturnDetail.fromJson(json as Map<String, dynamic>);
  }
}
